### 1 use Merton model
we can assume the 
equity is a long call option on V with strike F and
debt is a short put on V and long risk free F

In [13]:
from scipy.stats import norm
import numpy as np
alpha = 0.15 #(instantaneous expected asset return under the physical/objective measure)
sigma = 0.2 #(asset volatility)
r = 0.05# (constant, continuously np.exp(-d1**2)compounded risk-free rate)
A0=v0 =100 #(market value of the firm’s assets today, t = 0)
T =2 #(time to maturity for the bond)
F =80

### (a) set d1 and d2 and calculate the debt value D0 and Equity value E0

In [38]:
d1 = (np.log(v0/F) + (r + 0.5 * sigma ** 2) * T)/(sigma * T ** 0.5)
d2 = d1 - sigma * T ** 0.5
D0 = np.exp(-r * T) * F * norm.cdf(d2) + v0 * norm.cdf(-d1);D0

70.87977314490855

In [39]:
E0 = v0 * norm.cdf(d1) - np.exp(- r * T)* F * norm.cdf(d2);E0

29.120226855091445

#### use above parameter and black schole model calculate the $D0 = 70.8797$ and $E0=29.1202$

### (b) Calculate the vega

In [40]:
ve = v0 * norm.pdf(d1) * T **0.5;ve

24.744357919569442

In [41]:
vd = -v0 * norm.pdf(d1) * T ** 0.5 ;vd

-24.744357919569442

In [42]:
v_combine = D0 * vd + E0 * ve; v_combine

-1033.3131599540625

#### take derivative and calculate the $vega = \frac{\delta C}{\delta \sigma} = \frac{\delta p}{\delta \sigma}$
#### vega of equity = vega of call option = $24.744357919569442$
#### vega of equity = vega of F - vega of put option = $-24.744357919569442$<br/>
#### together the vega = $-1033.3131$<br/>

#### Since the vega is negative at this point, the minize the sigma will increase the value of assets 
#### and the value of the company. Thus, it is very essential to exercise the risk management.

### 2 
### (a) we can see the debt as 
### $D(T)$ = long call at K and short call at F and long K call binary at F
### =$max(A - k,0) - max(A - F, 0 ) + K \times 1_{A<F}$

#### The bankruptcy has nothing to do with equity holders in theory, thus the cost will not affect them anymore

In [32]:
E0 = v0 * norm.cdf(d1) - np.exp(- r * T)* F * norm.cdf(d2);E0

29.120226855091445

In [43]:
# set parameter for merton model
k = 30
d1_k =  (np.log(v0/k) + (r + 0.5 * sigma ** 2) * T)/(sigma * T ** 0.5)
d2_k = d1_k - sigma * T ** 0.5
d1_f = (np.log(v0/F) + (r + 0.5 * sigma ** 2) * T)/(sigma * T ** 0.5)
d2_f = d1 - sigma * T ** 0.5

In [44]:
# call at K use d1_k and d2_k
call_k = v0 * norm.cdf(d1_k) - np.exp(- r * T)* k * norm.cdf(d2_k)
# call at F use d1_f and d2_f
call_f = v0 * norm.cdf(d1_f) - np.exp(- r * T)* F * norm.cdf(d2_f)
# call binary at F use d2_f
call_binary = k * np.exp(- r * T)* norm.cdf(d2_f)

In [45]:
D0 = call_k - call_f + call_binary;D0

66.58004523638645

#### use above parameter and black schole model calculate the $D0 = 66.580045$ and $E0=29.1202$

### (b)

In [ ]:
# vega of call at K use d1_k and d2_k
v_k = v0 * norm.pdf(d1_k) * T **0.5
# vega of call at F use d1_f and d2_f
v_f = v0 * norm.pdf(d1_f) * T **0.5
# vega of call binary at F use d2_f
v_binary = k * norm.pdf(d2_f) * T **0.5

In [55]:
from sympy import *
import numpy as np
sigma = Symbol('x')

d1_f = (np.log(v0/F) + (r + 0.5 * sigma ** 2) * T)/(sigma * T ** 0.5)
d2_f = d1_f - sigma * T ** 0.5

d2_f_prime = d2_f.diff(sigma)

In [58]:
d2_f_prime

-2.22044604925031e-16 - 0.707106781186547*(x**2 + 0.32314355131421)/x**2